### ISSUE TITLE
Runtime slowdown when repeatedly fitting Prophet model in python

### ISSUE THREAD

**Issue by joaopcoelho** [facebookincubator/prophet#166]
Using Python, I noticed that when running a loop that fits a Prophet model in each iteration, each iteration seems to take an increasingly long time.

This happens even if the only thing we're doing is fitting the same data multiple times, which I expected would take about the same time. However, this same action gets about 10x slower after 300 iterations, and about 100x slower after 500 iterations.

I boiled it down to a minimal reproducible example. Here's the code and the results for 100, 200, 300 and 500 loop iterations (xaxis=iteration, yaxis=time(seconds)):

```
from fbprophet import Prophet
from time import time
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# create dummy df
dfsize = 100
dates = pd.date_range(start="20160101", periods=dfsize)
data = np.random.randn(dfsize)

df_prophet = pd.DataFrame()
df_prophet["ds"] = dates
df_prophet["y"] = data

output = {}

num_iterations = 100 #, 200, 300, 500
for i in xrange(num_iterations):
    print i

    t0 = time()

    m = Prophet()
    m.fit(df_prophet)

    t1 = time()
    
    output[i] = t1-t0

plt.scatter(output.keys(), output.values(), marker="o")
plt.show()

```
![figure_1_loop_100](https://cloud.githubusercontent.com/assets/10145707/25005187/2459900e-2058-11e7-8e19-7c62dc8280cd.png)
![figure_1_loop_200](https://cloud.githubusercontent.com/assets/10145707/25005185/24566fa0-2058-11e7-902e-3d7a94b0f502.png)
![figure_1_loop_300](https://cloud.githubusercontent.com/assets/10145707/25005186/245896b8-2058-11e7-8c7d-3399a9c23e60.png)
![figure_1_loop500](https://cloud.githubusercontent.com/assets/10145707/25005184/24562194-2058-11e7-83cc-2bf8a933975f.png)

I checked in R and the problem seems to go away - runtimes are constant and similar to what we get in Python on the first iterations. Here's the code and the result for 500 loop iterations (xaxis=time(s), yaxis=iteration):

```
library(prophet)

dates = seq(as.Date("2016/01/01"), by = "day", length.out = 100)
data = rnorm(100)
df = data.frame(dates, data)
colnames(df) = c("ds", "y")

dfsize = 100


niters = 500
time = c()
for (i in 1:niters){
    print(i)
    t0 = Sys.time()
    
    m <- prophet(df)

    t1 = Sys.time()
    
    time = c(time, t1-t0)
}

plot(time)
```

![loop_check_r_500](https://cloud.githubusercontent.com/assets/10145707/25005483/6de467de-2059-11e7-84d7-44499a4a5d3b.png)

Going out on a limb here, I think it might be related to #93 . However the problem here seems to be not the predict function (since it's not being used), but the fit (since it's the only thing being used).

Any help in figuring out why this happens / how to solve it is greatly appreciated. Thanks!

**Comment by Mathemagically** [comment]
@joaopcoelho Are you using the latest version or prophet provided yesterday, including [c164367](https://github.com/facebookincubator/prophet/commit/c164367c0841c2539e4aef30fbc986a36d465c4e)? If so, the problem noticed by you may be caused by the added garbage collection, which seems to slow down execution significantly.

The quick hack that I posted under [#93](https://github.com/facebookincubator/prophet/issues/93), adapting prophet and fit.prophet in a way that allows you to pass the stan model to these routines after loading it once, does not come with this overhead.

**Comment by joaopcoelho** [comment]
Hi @Mathemagically, thanks for your help! I'm using a version a few commits behind the latest one, so I'm not using that fix. I'll update to the latest version and see what happens. I'll also look into the hack you posted under #93.

**Comment by bletham** [comment]
@joaopcoelho Thanks for the great repro steps. #93 seems to be R specific, I'd expect this to be something different. Oddly, I actually don't see the same slowdown when I run exactly that script with 500 iterations!
![python_loop](https://cloud.githubusercontent.com/assets/6339760/25031389/75cea4e6-2081-11e7-883e-693cb3a2e9d2.png)

I tried this on both the pip version and the current Github version and did not experience slowdown on either. Maybe something to do with a package version? My pystan is 2.14.0.0, what's yours?

**Comment by bletham** [comment]
I suspect this issue is related to #162, and the Stan model being re-loaded at every iteration in the loop. https://github.com/facebookincubator/prophet/commit/5971a2369be1fc76af96c8ab6de1747c835acc95 fixes this, and so I hope will fix this issue too. Could you install from Github and check?

**Comment by joaopcoelho** [comment]
Hi @bletham, thanks for your help! My pystan version is the same, 2.14.0.0. I tried installing from Github and it solves the problem :) must have been the 5971a23 fix. 

For completeness I also tried installing the pip version and I with that the problem persists. So I'm not sure why you didn't experience slowdown when you did the same. In any case, it's solved with 5971a23. Thanks!

**Comment by shimosMA** [comment]
Hi all,
I'm new to GitHub as well as Prophet so I hope my question is in the correct place - 
I'm experiencing the same issue. I run a for-loop that fit and predicts a prophet model, and the running time increases dramatically from iteration to iteration. 
I installed the latest pystan and prophet packages, and the problem persists.
What can I do to solve this issue? 
I read through some of the suggestions made by @bletham but I couldn't make any breakthrough.
Many thanks in advance,
Matan
